File: converter/java/src/main/java/com/mlt/decoder/MltDecoder.java
package com.mlt.decoder;

import com.mlt.converter.encodings.PhysicalLevelTechnique;
import com.mlt.converter.encodings.StreamMetadata;
import com.mlt.data.Feature;
import com.mlt.data.Layer;
import com.mlt.data.MapLibreTile;
import com.mlt.metadata.MaplibreTileMetadata;
import me.lemire.integercompression.IntWrapper;
import org.locationtech.jts.geom.Geometry;
import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

public class MltDecoder {
    private MltDecoder(){}

    public static MapLibreTile decodeMlTile(byte[] tile, MaplibreTileMetadata.TileMetadata tileMetadata) throws IOException {
        var offset = new IntWrapper(0);
        var mltLayers = new ArrayList<Layer>();
        while(offset.get() < tile.length){
            List<Long> ids = null;
            Geometry[] geometries = null;
            var properties = new HashMap<String, List<Object>>();

            var layerVersion = tile[offset.get()];
            offset.increment();
            var layerInfos = DecodingUtils.decodeVarint(tile, offset, 4);
            var layerId = layerInfos[0];
            var tileExtent = layerInfos[1];
            var maxTileExtent = layerInfos[2];
            var numFeatures = layerInfos[3];

            var metadata = tileMetadata.getLayers(layerId);
            for(var columnMetadata : metadata.getFieldsList()){
                var columnName = columnMetadata.getName();
                var numStreams = DecodingUtils.decodeVarint(tile, offset, 1)[0];
                //TODO: compare based on ids
                if(columnName.equals("id")){
                    if(numStreams == 2){
                        var presentStreamMetadata = StreamMetadata.decode(tile, offset);
                        var presentStream = DecodingUtils.decodeBooleanRle(tile, presentStreamMetadata.numValues(), presentStreamMetadata.byteLength(), offset);
                    }

                    var idDataStreamMetadata = StreamMetadata.decode(tile, offset);
                    ids = idDataStreamMetadata.physicalLevelTechnique() == PhysicalLevelTechnique.FAST_PFOR?
                            IntegerDecoder.decodeIntStream(tile, offset, idDataStreamMetadata, false).
                                    stream().mapToLong(i -> i).boxed().collect(Collectors.toList()):
                            IntegerDecoder.decodeLongStream(tile, offset, idDataStreamMetadata, false);
                }
                else if(columnName.equals("geometry")){
                    System.out.println(layerId);
                    var geometryColumn = GeometryDecoder.decodeGeometryColumn(tile, numStreams, offset);
                    geometries = GeometryDecoder.decodeGeometry(geometryColumn);
                }
                else{
                        var propertyColumn = PropertyDecoder.decodePropertyColumn(tile, offset, columnMetadata, numStreams);
                        if(propertyColumn instanceof HashMap<?,?>){
                            var p = ((Map<String, Object>)propertyColumn);
                            for(var a : p.entrySet()){
                                properties.put(a.getKey(), (List<Object>)a.getValue());
                            }
                        }
                        else{
                            properties.put(columnName, (ArrayList)propertyColumn);
                        }
                }
            }

            var features = new ArrayList<Feature>(numFeatures);
            for(var j = 0; j < numFeatures; j++){
                var p = new HashMap<String, Object>();
                for(var propertyColumn : properties.entrySet()){
                    if(propertyColumn.getValue() == null){
                        p.put(propertyColumn.getKey(), null);
                    }
                    else{
                        var v = propertyColumn.getValue().get(j);
                        p.put(propertyColumn.getKey(), v);
                    }
                }
                var feature = new Feature(ids.get(j), geometries[j], p);
                features.add(feature);
            }
            var layer = new Layer(metadata.getName(), features);
            mltLayers.add(layer);
        }

        return new MapLibreTile(mltLayers);
    }

}

File: js/src/decoder/MltDecoder.ts
import { Feature } from '../data/Feature';
import { Layer } from '../data/Layer';
import { MapLibreTile } from '../data/MapLibreTile';
import { StreamMetadataDecoder } from '../metadata/stream/StreamMetadataDecoder';
import { TileSetMetadata } from "../metadata/mlt_tileset_metadata_pb";
import { IntWrapper } from './IntWrapper';
import { DecodingUtils } from './DecodingUtils';
import { IntegerDecoder } from './IntegerDecoder';
import { GeometryDecoder } from './GeometryDecoder';
import { PropertyDecoder } from './PropertyDecoder';
import { ScalarType } from "../metadata/mlt_tileset_metadata_pb";

export class MltDecoder {
    public static getTableMeta(tilesetMetadata: TileSetMetadata) {
        const tableMeta = [];
        for (let i=0; i < tilesetMetadata.featureTables.length; i++) {
            const featureTable = tilesetMetadata.featureTables[i];
            const types = [];
            for (const column of featureTable.columns) {
                const scalarColumn = column.type.value;
                if (scalarColumn !== undefined) {
                    types.push({
                        "name": String(column.name),
                        "type": Number(column.type.value.type.value)
                        });
                }
            }
            tableMeta[i] = {
                columns: types,
                name: featureTable.name
            };
        }
        return tableMeta;
    }

    public static decodeMlTile(tile: Uint8Array, tableMeta: any): MapLibreTile {
        const offset = new IntWrapper(0);
        const mltile = new MapLibreTile();
        while (offset.get() < tile.length) {
            let ids : BigInt64Array | Int32Array;
            let geometries;
            const properties = {};

            offset.increment();
            const infos = DecodingUtils.decodeVarint(tile, offset, 4);
            const version = tile[offset.get()];
            const extent = infos[1];
            const featureTableId = infos[0];
            const numFeatures = infos[3];
            const metadata = tableMeta[featureTableId];
            if (!metadata) {
                console.log(`could not find metadata for feature table id: ${featureTableId}`);
                mltile;
            }
            for (const columnMetadata of metadata.columns) {
                const columnName = columnMetadata.name;
                const numStreams = DecodingUtils.decodeVarint(tile, offset, 1)[0];
                if (columnName === "id") {
                    if (numStreams === 2) {
                        const presentStreamMetadata = StreamMetadataDecoder.decode(tile, offset);
                        // TODO: the return value of this function is not used, so advance offset without decoding?
                        DecodingUtils.decodeBooleanRle(tile, presentStreamMetadata.numValues(), offset);
                    } else {
                        throw new Error("Unsupported number of streams for ID column: " + numStreams);
                    }
                    const idDataStreamMetadata = StreamMetadataDecoder.decode(tile, offset);
                    const physicalType = columnMetadata.type;
                    if (physicalType === ScalarType.UINT_32) {
                        ids = IntegerDecoder.decodeIntStream(tile, offset, idDataStreamMetadata, false);
                    } else if (physicalType === ScalarType.UINT_64){
                        ids = IntegerDecoder.decodeLongStream(tile, offset, idDataStreamMetadata, false);
                    } else {
                        throw new Error("Unsupported ID column type: " + physicalType);
                    }
                } else if (columnName === "geometry") {
                    const geometryColumn = GeometryDecoder.decodeGeometryColumn(tile, numStreams, offset);
                    geometries = GeometryDecoder.decodeGeometry(geometryColumn);
                } else {
                    const propertyColumn = PropertyDecoder.decodePropertyColumn(tile, offset, columnMetadata.type, numStreams);
                    if (propertyColumn instanceof Map) {
                        throw new Error("Nested properties are not implemented yet");
                    } else {
                        properties[columnName] = propertyColumn;
                    }
                }
            }
            mltile.layers[metadata.name] = MltDecoder.convertToLayer(ids, extent, version, geometries, properties, metadata.name, numFeatures);
        }

        return mltile;
    }

    private static convertToLayer(ids: BigInt64Array | Int32Array, extent, version, geometries, properties, name : string, numFeatures: number): Layer {
        const features: Feature[] = new Array(numFeatures);
        const vals = Object.entries(properties);
        for (let j = 0; j < numFeatures; j++) {
            /* eslint-disable @typescript-eslint/no-explicit-any */
            const p: { [key: string]: any } = {};
            for (const [key, value] of vals) {
                if (value) {
                    const val = value[j];
                    if (val) {
                        p[key] = val;
                    }
                }
            }
            features[j] = new Feature(ids[j], extent, geometries[j], p)
        }

        return new Layer(name, version, features);
    }
}
